app/controllers/session_controller.rb

```diff
@@ -612,4 +642,49 @@ def render_sso_error(status:, text:)
   def sso_url(sso)
     sso.to_url
   end
+
+  # the invite_key will be present if set in InvitesController
+  # when the user visits an /invites/xxxx link; however we do
+  # not want to complete the SSO process of creating a user
+  # and redeeming the invite if the invite is not redeemable or
+  # for the wrong user
+  def validate_invitiation!(sso)
+    invite_key = secure_session["invite-key"]
+    return if invite_key.blank?
+
+    @invite = Invite.find_by(invite_key: invite_key)
+
+    if @invite.blank?
+      raise Invite::ValidationFailed.new(I18n.t("invite.not_found", base_url: Discourse.base_url))
+    end
+
+    if @invite.redeemable?
+      if !@invite.is_invite_link? && sso.email != @invite.email
+        raise Invite::ValidationFailed.new(I18n.t("invite.not_matching_email"))
+      end
+    else
+      if @invite.expired?
+        raise Invite::ValidationFailed.new(I18n.t('invite.expired', base_url: Discourse.base_url))
+      elsif @invite.redeemed?
+        raise Invite::ValidationFailed.new(I18n.t('invite.not_found_template', site_name: SiteSetting.title, base_url: Discourse.base_url))
+      end
+    end
+  end
+
+  def redeem_invitation(sso)
+    InviteRedeemer.new(
+      invite: @invite,
+      username: sso.username,
+      name: sso.name,
+      ip_address: request.remote_ip,
+      session: session,
+      email: sso.email
```

Comment:
Do we need to pass this in if it is an email invite link? The email can already be retrieved via the `invite` object.

Reply:
Yes it does, the email passed in is used for comparison with invite.email

https://github.com/discourse/discourse/blob/956e3594ada1dfdea62c81ee070a15dbf1bbe7d4/app/models/invite_redeemer.rb#L79-L79

Reply:
ahhh icic